PPTT-2696 added ecommerce guide and labels to ecommerce functions (#631)
PiwikPRO/PPMS-PublicDocs: data_collection/web/guides.rst

----------------------

Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop p

E-commerce
----------

JavaScript API supports 3 types of e-commerce interactions: :ref:`Category and product views<guide_tracking_category_and_product_views>`, :ref:`Cart updates<guide_tracking_cart_updates>` and :ref:`Orders<guide_tracking_orders>`.

.. _guide_tracking_category_and_product_views:

Tracking category and product views
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Usually, the first e-commerce-related action a user performs on a website is browsing products. :ref:`setEcommerceView<js-api-setEcommerceView>` method allows us to track both category views and product views.

> To track a category view, use ``setEcommerceView`` method **before** tracking the page view, like this:

Reviewer: ```suggestion
This is added. To track a this is changed view, use ``setEcommerceView`` method **before** tracking the page view, like this:
```

.. code:: javascript

    // set category to "Smartphones"
    _paq.push(["setEcommerceView", false, false, "Smartphones"]);

    // track page view
    _paq.push(["trackPageView"]);

The same method can be used for tracking product views. Again, it must be called **before** tracking a page view. Example:

.. code:: javascript

    // set product to "SUPER Phone A40 White"
    _paq.push(["setEcommerceView", "71253029", "SUPER Phone A40 White", "Smartphones", 1499.99]);

    // track page view
    _paq.push(["trackPageView"]);

The snippet above tracks product view of a smartphone with

* SKU *71253029*
* name *SUPER Phone A40 White*
* category *Smartphones*

> * price *1499.99*

Author: I've added description to the code blocks.

    // track page view
    _paq.push(["trackPageView"]);

.. _guide_tracking_cart_updates:

Tracking cart updates
^^^^^^^^^^^^^^^^^^^^^

Another type of e-commerce activity one can track is updating a shopping cart. With it, we are able to measure how often users don't complete the ordering process and what products stay in abandoned carts.

Tracking a cart update has two steps: registering items from the cart and sending them. The following example uses two methods - :ref:`addEcommerceItem<js-api-addEcommerceItem>` and :ref:`trackEcommerceCartUpdate<js-api-trackecommercecartupdate>` - to achieve exactly that.

.. code:: javascript

    // user added one chocolate bar to an empty shopping cart

    // register chocolate bar
    _paq.push(["addEcommerceItem", "82775027", "MEGA Milk Chocolate 200g", "Candy", 6.00, 1]);

    // track cart update
    _paq.push(["trackEcommerceCartUpdate", 6.00]);

This code snippet sends a cart update event with a cart containing one item (SKU *candy-12837*, name *MEGA Milk Chocolate 200g*, category *Candy*, price *6.00*) and having total value of *6.00*.

> The list of registered items **is not persisted** in the browser. If a user adds another item to the cart and the page reloads in between cart updates, the items that were already registered must be registered again.

Reviewer 2: I'd just write that it's stateless. Every IT person should understand that.
It's important because it means that you have to send full content of the cart every time it's updated. It also means that of something is removed from the cart, then you have to send whole current content of the cart, just without this one item.
Maybe you should rephrase your description so it reflects the situation better.

```suggestion
The list of registered items **stateless**. If a visitor adds another item to the cart and the page reloads in between cart updates, the items that were already registered must be registered again.
```

Author: It's not completely stateless, it remembers what is in the cart until you reload a page, in case of a SPA that might be never. I'll try to formulate it differently.

    </script>

E-commerce
----------

JavaScript API supports 3 types of e-commerce interactions: :ref:`Category and product views<guide_tracking_category_and_product_views>`, :ref:`Cart updates<guide_tracking_cart_updates>` and :ref:`Orders<guide_tracking_orders>`.

.. _guide_tracking_category_and_product_views:

Tracking category and product views
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Usually, the first e-commerce-related action a visitor performs on a website is browsing products. :ref:`setEcommerceView<jtc-api-setEcommerceView>` method allows us to track both category views and product views.

To track a category view, use :ref:`setEcommerceView<jtc-api-setEcommerceView>` method **before** tracking the page view, like this::

    // set category to "Smartphones"
    _paq.push(["setEcommerceView", false, false, "Smartphones"]);

    // track page view
    _paq.push(["trackPageView"]);

The same method can be used for tracking product views. Again, it must be called **before** tracking a page view. Example::

>     // set product

Reviewer 2: Why didn't you use code styling block like `.. code-block:: javascript` ?

Reviewer 2: There are a couple of more code blocks without code styling - if possible please use `.. code-block:: javascript` for them too.

Author: This is a simplified syntax for code blocks. You can choose the language globally for a document with `.. highlight:: js` directive. It was missing here though. Thanks for catching that.
https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks

Reviewer 2: Personally I'd prefer to have it literally declared with every code block - maybe that's a little bit more to write but is on the other hand is more clear what is going on.
But that's my personal opinion - you do it as you judgment tells you.

    _paq.push(["setEcommerceView",
        "71253029",              // SKU (stock-keeping unit)
        "SUPER Phone A40 White", // name
        "Smartphones",           // category
        1499.99                  // price
    ]);

    // track page view
    _paq.push(["trackPageView"]);

``category`` parameter of the :ref:`setEcommerceView<jtc-api-setEcommerceView>` method accepts not only string values, but also arrays of strings. This is useful for tracking products that belong to more than one category, or tracking pages that list products from multiple categories. ::

    // set product to "SUPER Watch B20 Silver" belonging to 2 categories
    _paq.push(["setEcommerceView", "00492710", "SUPER Watch B20 Silver", ["New offer", "Smartwatches"], 700.00]);

    // track page view
    _paq.push(["trackPageView"]);

.. _guide_tracking_cart_updates:

Tracking cart updates
^^^^^^^^^^^^^^^^^^^^^

> Another type of e-commerce activity you can track is an update to a shopping cart. With it, we are able to measure how often visitors don't complete the ordering process and what products stay in abandoned carts.

Reviewer 2: ```suggestion
Another type of e-commerce activity you can track is an update of a shopping cart. With it, we are able to measure how often visitors don't complete the ordering process and what products stay in abandoned carts.
```

    // visitor added one chocolate bar to an empty shopping cart

    // register chocolate bar
    _paq.push(["addEcommerceItem", "82775027", "MEGA Milk Chocolate 200g", "Candy", 6.00, 1]);

    // track cart update
    _paq.push(["trackEcommerceCartUpdate", 6.00]);

This code snippet sends a cart update event with a cart containing one item (SKU *candy-12837*, name *MEGA Milk Chocolate 200g*, category *Candy*, price *6.00*) and having total value of *6.00*.

The list of registered items is stored only in memory. **Reloading the page will clear the list** and the previously registered items will have to be added again. ::

    // visitor added one mango fruit to a shopping cart with one chocolate bar

    // register previously added items
    _paq.push(["addEcommerceItem", "82775027", "MEGA Milk Chocolate 200g", "Candy", 6.00, 1]);

    // register the new item
    _paq.push(["addEcommerceItem", "01809926", "FRUTASTIC Mango", "Fruits & vegetables", 4.00, 1]);

    // track cart update

>     _paq.push(["trackEcommerceCartUpdate", 10.00]);

Reviewer 2: Could you describe the parameters as the comment in the code the same way as you described it at the first occurrence of the `setEcommerceView` ?

> If you are not sure what items have been registered, use :ref:`getEcommerceCart<jtc-api-getEcommerceItems>` method. ::

Reviewer 2: Would be nice to wrap this debugging hint in the `info` block.
